File: solutions/python/variable-length-quantity/1/variable_length_quantity.py
EIGHT_BIT_MASK = 0x80
SEVEN_BIT_MASK = 0x7F


def encode_int(n: int) -> list[int]:
    byte_lst = [n & SEVEN_BIT_MASK]
    n >>= 7
    while n > 0:
        byte_lst.append(n & SEVEN_BIT_MASK | EIGHT_BIT_MASK)
        n >>= 7
    return byte_lst[::-1]


def encode(numbers: list[int]) -> list[int]:
    byte_lst = []
    for n in numbers:
        byte_lst.extend(encode_int(n))
    return byte_lst


def decode(bytes_: list[int]) -> list[int]:
    values = []
    number = 0
    for index, byte in enumerate(bytes_):
        number = (number << 7) + (byte & SEVEN_BIT_MASK)
        if not byte & EIGHT_BIT_MASK:
            values.append(number)
            number = 0
        elif index == len(bytes_) - 1:
            raise ValueError("incomplete sequence")
    return values

File: solutions/javascript/variable-length-quantity/1/variable-length-quantity.js
const LENGTH = 7;
const CONT_BITS = 1 << LENGTH;
const DATA_BITS = CONT_BITS - 1;

/**
 * Encodes a single value using variable-length quantity encoding.
 *
 * @param {number} val - The value to encode.
 * @returns {number[]} - The encoded value as an array of bytes.
 */
const encodeOne = (val) => {
  const buf = [];
  let left = val;

  while (left) {
    const bits = (left & DATA_BITS) | CONT_BITS; // set continuation everywhere
    left = left >>> LENGTH;
    buf.push(bits);
  }
  buf[0] = buf[0] & DATA_BITS; // cancel the last continuation
  return buf.reverse();
};

/**
 * Decodes a single value from variable-length quantity encoding.
 *
 * @param {number[]} buf - The encoded value as an array of bytes.
 * @returns {number} - The decoded value.
 */
const decodeOne = (buf) => {
  let val = 0;

  for (const byte of buf) {
    val = (val << LENGTH) | (byte & DATA_BITS);
  }
  return val >>> 0; // convert to unsigned 32-bit
};

/**
 * Encodes an array of values using variable-length quantity encoding.
 *
 * @param {number[]} data - The array of values to encode.
 * @returns {number[]} - The encoded values as an array of bytes.
 */
export const encode = (data) => {
  return data.reduce((buf, val) => buf.concat(encodeOne(val)), []);
};

/**
 * Decodes an array of bytes from variable-length quantity encoding.
 *
 * @param {number[]} data - The encoded values as an array of bytes.
 * @returns {number[]} - The decoded values.
 * @throws {Error} - If the sequence is incomplete.
 */
export const decode = (data) => {
  let start = 0;
  const vals = [];

  for (let i = 0; i < data.length; i++) {
    if (~data[i] & CONT_BITS) {
      vals.push(decodeOne(data.slice(start, i + 1)));
      start = i + 1;
    }
  }
  if (start < data.length) {
    throw new Error('Incomplete sequence');
  }
  return vals;
};
